### PR TITLE
Adding PFClusterIsolation to reco Egamma objects

### PR DESCRIPTION
**Pull request by Sam-Harper**
PFCluster isolation is used in the HLT and is becoming more popular in E/gamma as useful isolation.

It is already stored in AOD via value maps and is embedded into pat::Electrons and pat::Photons. This change puts it directly in the reco::GsfElectron and reco::Photon for ease of use.  It should be CPU and diskspace neutral (if anything, slightly less diskspace as it replaces valuemaps) as its simply changing where the isolation value is stored.

The value maps are now added to the electron,photon during the finalisation step in the same way as PFIsolations are. As such the event content is modified to drop these ValueMaps as they are unnecessary (and keyed to the tmp versions of the electrons/photons). The reducedEgamma and pat configs were updated to not to store the PF cluster isolation.  However they still have a capability to read in value maps and set the PFCluster isolation. This is done incase we wish to use this release to re-miniAOD 8X or 9X samples where a customisation function would configure these modules to read the value maps.  Note, I wasnt sure how to do such a customisation function so any advice here would be welcome. 

The only slightly tricky part was with the OOT photons which dont have PF isolation and so dont have a finalisation step. I added one for them and they should be untouched otherwise. @kmcdermo FYI

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaPhotonProducers/python/ootPhotons_cff.py [hunk 1]
```diff
@@ -1,32 +1,48 @@
...
     candidateP4type = "fromEcalEnergy",
     reconstructionStep = "oot",
     pfEgammaCandidates = "",
     valueMapPhotons = ""
     )
+del ootPhotonsTmp.regressionConfig
+
+ootPhotons = _gedPhotons.clone(
+    photonProducer = 'ootPhotonsTmp',
+    candidateP4type = "fromEcalEnergy",
```
Comment (kmcdermo) [inline RecoEgamma/EgammaPhotonProducers/python/ootPhotons_cff.py]: @Sam-Harper , checking https://github.com/Sam-Harper/cmssw/blob/6e6a25fd65e53bbf5a43d99fecab8a2f46712896/RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc#L841-L843, this means that the regression is effectively skipped in the final step for ootPhotons, correct?  

I guess the nice thing here is that, now the symmetry between ged and oot photons is closer with a tmp and final step for both, even if nothing really happens in the final step for oot photons.

Comment (Sam-Harper) [reply RecoEgamma/EgammaPhotonProducers/python/ootPhotons_cff.py]: Yes, the intention was to skip the regression for OOT photons.

**Comment by Sam-Harper** [comment]
Okay I fixed the problems, sorry about that. However on both RAL and lxplus I simply cant access the input files for the 136.7611 workflow to confirm it worked.  The HION workflow is confirmed to work now. 

I now should have the legacy re-miniAOD 80X working correctly, the modifers are now properly updated.  A question @slava77, this code wont allow re-miniAOD of 9X samples in 10X so I would like a modifier for this ala re-miniAOD 80X. Could I create such a modifier, something like Configuration/Eras/python/Modifier_run2_miniAOD_9XLegacy_cff.py with a modifier called run2_miniAOD_9XLegacy = cms.Modifier() ?

**Comment by slava77** [comment]
On 1/9/18 6:27 AM, Sam-Harper wrote:
> Okay I fixed the problems, sorry about that. However on both RAL and
> lxplus I simply cant access the input files for the 136.7611 workflow to
> confirm it worked. The HION workflow is confirmed to work now.
> 
> I now should have the legacy re-miniAOD 80X working correctly, the
> modifers are now properly updated. A question @slava77
> <https://github.com/slava77>, this code wont allow re-miniAOD of 9X
> samples in 10X so I would like a modifier for this ala re-miniAOD 80X.
> Could I create such a modifier, something like
> Configuration/Eras/python/Modifier_run2_miniAOD_9XLegacy_cff.py with a
> modifier called run2_miniAOD_9XLegacy = cms.Modifier() ?

Is it affecting 92X and 94X similarly?

If so, Modifier_run2_miniAOD_9X_cff should be OK, no need to add "Legacy"

**Comment by Sam-Harper** [comment]
Yes. The issue is that in 92X + 94X (and 80X)  recoed samples, the PFcluster isolation is stored in a value map so when you make the miniAOD you read in those value maps and store them in the pat::Electron/Photon. This PR does away with this and stores them in the reco::GsfElectron/reco::Photon.

So to reminiAOD 9X samples in 10X, you need the modifier to tell pat to read the values from value maps as was done before. 

Eg, basically this again but with 9X modifier in its place: https://github.com/Sam-Harper/cmssw/blob/29ee74c7ef4b5234bbf54bb020f617d0cfd90672/RecoEgamma/EgammaPhotonProducers/python/reducedEgamma_cfi.py#L61-L86 
It'll be slightly different due to the OOT photons.

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
On 1/9/18 6:27 AM, Sam-Harper wrote:
> Okay I fixed the problems, sorry about that. However on both RAL and
> lxplus I simply cant access the input files for the 136.7611 workflow to
> confirm it worked. The HION workflow is confirmed to work now.
> 

from jenkins logs you can see that the file consumed is
root://eoscms.cern.ch//eos/cms/store/user/cmsbuild/store/data/Run2016E/JetHT/AOD/18Apr2017-v1/100000/02FB8769-5433-E711-BBDE-0025905A6076.root

you can try to use it directly

**Comment by Sam-Harper** [comment]
Okay it should all be fixed now. Some fun with ootPhotons. I have also confirmed that the miniAOD objects pf cluster isolation is untouched in the re-miniAOD workflow. 

I added a Modifier for running re-miniAOD on 9X produced AOD. It doesnt work because tau modules crash (ie they didnt have this idea). But E/gamma should work.

**Comment by perrotta** [comment]
please test

File: RecoEgamma/EgammaElectronProducers/plugins/GEDGsfElectronFinalizer.cc [hunk 2]
```diff
@@ -103,6 +105,12 @@ void GEDGsfElectronFinalizer::produce( edm::Event & event, const edm::EventSetup
      isoVariables.sumPhotonEt = (*(isolationValueMaps)[1])[myElectronRef];
      isoVariables.sumNeutralHadronEt = (*(isolationValueMaps)[2])[myElectronRef];
      isoVariables.sumPUPt = (*(isolationValueMaps)[3])[myElectronRef];
+     if(isolationValueMaps.size()<=4){
```
Comment (perrotta) [inline RecoEgamma/EgammaElectronProducers/plugins/GEDGsfElectronFinalizer.cc]: How is this different from checking ```nDeps_``` ?

Comment (Sam-Harper) [reply RecoEgamma/EgammaElectronProducers/plugins/GEDGsfElectronFinalizer.cc]: My apologies, this was some debugging code left in accidentally, its completely unnecessary and should be removed. I will do this shortly.

**Comment by davidlange6** [comment]
please test

**Comment by Sam-Harper** [comment]
my bad (sorry I really have been shocking with these two PRs), when I reset, I forgot to re-add a file, fixed now

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please test
(just to kep test results updated here)

**Comment by Sam-Harper** [comment]
Ah on this, is there a reason this didnt get merged? Or just an oversight in the very busy time of building 10_0_0?

**Comment by perrotta** [comment]
Sam-Harper <notifications@github.com> ha scritto:

> Ah on this, is there a reason this didnt get merged? Or just an  
> oversight in the very busy time of building 10_0_0?

Indeed, it was not listed among the priorities for 10_0_0...
Jumping on it now

**Comment by Sam-Harper** [comment]
@perrotta , okay no problem, while it would have been nice, completely understand it that there were bigger fish to fry. If we could get it merged relatively early in 10_1_X cycle, that would be great though. I have some more potential data format changes related to HCAL.

**Comment by perrotta** [comment]
Looking at the reco outputs, ecal/hcalPFClusterIso() are filled for 
- recoGsfElectrons::gedGsfElectrons, but not for recoGsfElectrons::uncleanedOnlyGsfElectrons
- recoPhotons:gedPhotons and recoPhotons:ootPhotons, but not for recoPhotons:photons

This is what expected, because non-ged electrons/photons do not have the ged finalization step.

I also verified that the event output sizes do not increase: in fact the reco output even tinily shrinks, thanks to the removal of the floatedmValueMap's (as anticipated in the PR description)

As soon as some conclusion is reached for the miniAOD workflows on top of 92X  (see https://github.com/cms-sw/cmssw/pull/21811#pullrequestreview-92640247 and  https://github.com/cms-sw/cmssw/pull/21811#issuecomment-361662328), this PR can be signed off by reco.

**Comment by slava77** [comment]
On 1/31/18 12:16 AM, perrotta wrote:
> Looking at the reco outputs, ecal/hcalPFClusterIso() are filled for
> 
>   * recoGsfElectrons::gedGsfElectrons, but not for
>     recoGsfElectrons::uncleanedOnlyGsfElectrons
>   * recoPhotons:gedPhotons and recoPhotons:ootPhotons, but not for
>     recoPhotons:photons
> 
> This is what expected, because non-ged electrons/photons do not have the
> ged finalization step.
> 
> I also verified that the event output sizes do not increase: in fact the
> reco output even tinily shrinks, thanks to the removal of the
> floatedmValueMap's (as anticipated in the PR description)
> 
> As soon as some conclusion is reached for the miniAOD workflows on top
> of 92X (see #21811 (review)
> <https://github.com/cms-sw/cmssw/pull/21811#pullrequestreview-92640247>
> and #21811 (comment)
> <https://github.com/cms-sw/cmssw/pull/21811#issuecomment-361662328>),
> this PR can be signed off by reco.
> 
> —

I propose to change the name as suggested  (run2_miniAOD_94XFall17 )and
then proceed with integration.

**Comment by Sam-Harper** [comment]
thanks Andrea and Slava, I've updated the name now

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
+1
- ecal/hcalPFClusterIso() are filled for ged electrons/photons
- output sizes do not increase: in fact the reco output even tinily shrinks, thanks to the removal of the floatedmValueMap's (as anticipated in the PR description)
- latest jenkins tests show 50 differences in the comparisons: this is simply due to the introduction of those ecal/hcalPFClusterIso() in the monitored quantities for miniAOD. No other differences are observed, as expected

**Comment by slava77** [comment]
@fabiocos 
Please clarify your plans for this PR.
This contains a new modifier run2_miniAOD_94XFall17 which will be needed for re-miniAOD PRs and backports. If this PR will take longer than until Monday, then perhaps the feature introducing the new modifier will need to be split off into a separate PR (assuming that it can go in quickly).

**Comment by fabiocos** [comment]
+operations

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**